### ISSUE TITLE
Support Log methods for submitting data (weight, fat, etc)

### ISSUE THREAD

**Issue by AlexGhiondea** [aarondcoleman/Fitbit.NET#44]
This is a great API! Thanks for putting it together!

I added the LogWeight method that I needed it for my project. I am happy to submit a PR with it.

I have a couple of questions before I get started:
- Which branch should the change go to?
- For logging methods I prefer having the explicit parameters in the signature and not the Log object (https://github.com/aarondcoleman/Fitbit.NET/blob/master/Fitbit/FitbitClient.cs#L715)
- There is some opportunity to refactor the code to reduce code duplication but I am not sure how that will conflict with the async work (see first bullet :) )

**Comment by WestDiscGolf** [comment]
@AlexGhiondea thanks for the feedback on the library.

As for the branch target, if it is targeting the existing fitbit client (xml based, uses restsharp) then please direct the PR to the master branch. 

The issue with posting the individual parameters instead of just the log object is if any additional properties are added/removed to/from the object then the signature will make a breaking change which is bad. Please keep the new methods consistent with the rest of the library, it will make the PR more likely to be accepted and make migration into the async world smoother.

With regards to the refactor which version are you referring to? If it's V1 then active development on that branch has been reduced to accepting PR and defect resolution. If it's V2 then this is still in active development and influx but it would be good to hear your suggestions and we can take it from there.

**Comment by AlexGhiondea** [comment]
I will send the PR against master as the change is to simply add 2 methods (one for Weight and one for Fat).

The concern I have with using parameter objects is that they are not as expressive. Yes, you can add additional parameters without changing the signature but they also make it hard to figure out which parameters are being used by the method they are passed into. Additionally, parameter objects are better suited to APIs that have a lot of parameters that need to be passed in (5+).

For the method to add the Weight, the class `WeightLog` has additional properties than the one I would like to add (which is `Weight`)
```csharp
    public class WeightLog
    {
        public long LogId { get; set; }
        public float Bmi { get; set; }
        public DateTime Date { get; set; }
        public DateTime Time { get; set; }
        public float Weight { get; set; }
        public DateTime DateTime { get { return Date.Date.Add(Time.TimeOfDay); } }
    }
```

As a consumer of the API it is hard to tell which of the properties on `WeightLog` I should pass and it might be frustrating when I set the `Bmi` but it does not get saved.

Part of the problem is that the same objects are used as both input and output to the method. If you want the parameter object then I would suggest creating new classes to define the input to the Log method. That way you decouple the input/output object and still have the flexibility you want. The downside of this approach is that you will end up with a lot of classes that have one or two properties

I understand your desire to maintain compatibility with existing APIs -- however there are ways you can ensure binary compatibility even if you go the overload route. One of them is to keep both the old and the new APIs in the surface area and make sure the old API provides a suitable default for the new parameter. The .NET Framework has done that in quite a few places.

The refactoring I had in mind was around the way the calls to the service are done. Instead of having each Log method setup the call I was going to move to capturing that logic in a single method that can be used by all the Log methods. 

I will take a look at the async branch to see if this is still applicable.

**Comment by aarondcoleman** [comment]
If it's a sync method and uses restsharp deserializing, submit the PR to my master branch. I totally get where you're coming form on weight logging. I wouldn't say we have to be hard and fast with this rule of an object because, as you say, the output model doesn't match the input. A few overloads won't kill anyone.

Also -- this is the V1 library that we'll eventually freeze and put in non-maintenance mode in favor of the one @WestDiscGolf  and I have been working on (see his master, my async-portable branch). They'll run side by side without collision and let you migrate. 

So basically, if it works, and it isn't too ugly -- it's in!
--Aaron

**Comment by AlexGhiondea** [comment]
Ping? :)
